### PR TITLE
enhancement(kubernetes_logs): preserve event ordering

### PR DESCRIPTION
**Pull request by sillent**
PR that may solve the problem in issue #10883

### CODE REVIEW

File: src/sources/kubernetes_logs/mod.rs [hunk 1]
```diff
@@ -445,10 +454,13 @@ impl Source {
 
         let mut stream = partial_events_merger.transform(Box::pin(events));
         let event_processing_loop = out.send_event_stream(&mut stream);
-
-        let mut lifecycle = Lifecycle::new();
+        let mut lc: Lifecycle;
+        match ordered {
+            true => lc = Lifecycle::new(lifecycle::LifecycleOrder::Ordered),
+            false => lc = Lifecycle::new(lifecycle::LifecycleOrder::Unordered),
```
Comment (bruceg) [inline src/sources/kubernetes_logs/mod.rs:460]: If you move the `ordered` parameter to `Lifecycle::new`, I think you can hide the `LifecycleOrder` enum entirely in that module and not have it exposed here. I would also recommend against shortening the name.

**Review by bruceg — CHANGES_REQUESTED** [review]
Thank you for your contribution, @sillent. It's not obvious to me that this approach will solve the problem in #10883, but I am not very familiar with the `kubernetes_logs` source so I'll leave that commentary to someone else. I did leave a couple of requests for improving the code patterns. Also, make sure to format your code before submitting by running `cargo fmt` if your editor doesn't do that automatically.

File: src/sources/kubernetes_logs/mod.rs [hunk 2]
```diff
@@ -138,6 +138,10 @@ pub struct Config {
     /// How long to delay removing entries from our map when we receive a deletion
     /// event from the watched stream.
     delay_deletion_ms: usize,
+
+    /// What's order the events from log files must be passed to the sinks
+    /// By default all events is unordered
```
Comment (spencergilbert) [inline src/sources/kubernetes_logs/mod.rs:143]:
```suggestion
    /// If events must be properly ordered when sending downstream,
    /// by default all events are unordered.
```

File: website/cue/reference/components/sources/kubernetes_logs.cue [hunk 3]
```diff
@@ -288,6 +288,13 @@ components: sources: kubernetes_logs: {
 				unit:    "milliseconds"
 			}
 		}
+		order: {
+			common: false
+			description: """
+				Specifies the order in which events are processed.
```
Comment (spencergilbert) [inline website/cue/reference/components/sources/kubernetes_logs.cue:294]:
```suggestion
				Specifies if the events are processed in order.
```

File: src/sources/kubernetes_logs/lifecycle.rs [hunk 4]
```diff
@@ -118,7 +164,11 @@ impl<'bound, 'lc> Slot<'bound, 'lc> {
     /// shutdown via the signal passed from the corresponding
     /// [`ShutdownHandle`].
     pub fn bind(self, future: BoxFuture<'bound, ()>) {
-        self.lifecycle.futs.push(future);
+        match self.lifecycle.order {
+            LifecycleOrder::Ordered => self.lifecycle.futs_ordered.push(future),
+            LifecycleOrder::Unordered => self.lifecycle.futs_unordered.push(future),
+        }
+        // self.lifecycle.futs.push(future);
```
Comment (spencergilbert) [inline src/sources/kubernetes_logs/lifecycle.rs:171]:
```suggestion
```

Old code left commented out

**Review by spencergilbert — COMMENTED** [review]
I left a couple of nits around docs, comments, and a stray old line. Similarly to @bruceg and @jszwedko, it's unclear to me how the lifecycle code is influencing the order in which events are processed by the underlying file server.

Given that with the patch, you've (manually?) tested and confirmed the behavior is corrected, I'm curious if we can include some sort of test that confirms, and prevents regressions, in the ordered processing of events here.

As an aside, I'm also wondering if this option would be better suited in the underlying file server that processes the log files - though I know we do plan on rewriting much of that component.

File: src/sources/kubernetes_logs/lifecycle.rs [hunk 5]
```diff
@@ -8,12 +8,18 @@ use futures::{
...
     FutureExt, StreamExt,
 };
 
 use crate::shutdown::{ShutdownSignal, ShutdownSignalToken};
 
+#[derive(Debug)]
+pub enum LifecycleOrder{
+    Ordered,
+    Unordered,
+}
```
[the remark this answers is not in the document]
Comment (sillent) [reply src/sources/kubernetes_logs/lifecycle.rs]: @bruceg thank you for comments. I've changed the logic. Now the ordering is passed to the new method of Lifecycle.

File: src/sources/kubernetes_logs/mod.rs (hunk 1 above)
Comment (sillent) [reply src/sources/kubernetes_logs/mod.rs]: @bruceg now into Lifecycle::new just passes an 'ordered' as arguments

File: src/sources/kubernetes_logs/mod.rs (hunk 2 above)
Comment (sillent) [reply src/sources/kubernetes_logs/mod.rs]: fixed

File: website/cue/reference/components/sources/kubernetes_logs.cue (hunk 3 above)
Comment (sillent) [reply website/cue/reference/components/sources/kubernetes_logs.cue]: fixed, and closed quotes added

File: src/sources/kubernetes_logs/lifecycle.rs (hunk 4 above)
Comment (sillent) [reply src/sources/kubernetes_logs/lifecycle.rs]: fixed

**Comment by sillent** [comment]
> I left a couple of nits around docs, comments, and a stray old line. Similarly to @bruceg and @jszwedko, it's unclear to me how the lifecycle code is influencing the order in which events are processed by the underlying file server.
> 
> Given that with the patch, you've (manually?) tested and confirmed the behavior is corrected, I'm curious if we can include some sort of test that confirms, and prevents regressions, in the ordered processing of events here.
> 
> As an aside, I'm also wondering if this option would be better suited in the underlying file server that processes the log files - though I know we do plan on rewriting much of that component.

@spencergilbert hello, thank you for you comment

I'm not sure but I think it's a problem with the order of events is starting here
`
  let mut lifecycle = Lifecycle::new();
        {
            let (slot, shutdown) = lifecycle.add();
            let fut = util::run_file_server(file_server, file_source_tx, shutdown, checkpointer)
                .map(|result| match result {
                    Ok(FileServerShutdown) => info!(message = "File server completed gracefully."),
                    Err(error) => emit!(KubernetesLifecycleError {
                        message: "File server exited with an error.",
                        error,
                    }),
                });
            slot.bind(Box::pin(fut));
        }
`
when the start a FileServer, that watching for new files and start reading it and generate new events.

The problem with FuturesUnordered is appear when the file fill very fast (application write log too fast), when the first file is filled up, and system rotating him, the Vector is still not fully processed him, and the new files appears, the 'FileServer' start reading the new log-file and start generating events. The events from first file is mixing up by the events from the new file, and in destination we saw a unordered lines.

**Comment by bruceg** [comment]
I think your explanation makes sense. Could you confirm that using a `FuturesOrdered` resolves the problem you are seeing. If so, is there any reason not to use it unconditionally?

**Comment by spencergilbert** [comment]
> I think your explanation makes sense. Could you confirm that using a `FuturesOrdered` resolves the problem you are seeing. If so, is there any reason not to use it unconditionally?

Yeah, that's reasonable - unless there's a noticeable performance impact it seems like this would be strictly "better" behavior.

**Comment by sillent** [comment]
> Yeah, that's reasonable - unless there's a noticeable performance impact it seems like this would be strictly "better" behavior.

In issue #10883 my colleague was testing the functionality with FuturesOrdered and FuturesUnordered, and with FuturesOrdered the message order is not violated, however with the same number of message with FuturesUnordered we have order breaking.
So if the FuturesOrdered is preferred for this situation, i'l do changed the FuturesUnordered to FuturesOrdered and is it enough?

**Comment by spencergilbert** [comment]
Thanks again @sillent - it sounds like just switching to `FuturesOrdered` is the best way forward, given it seems like a strictly better default.

**Comment by sillent** [comment]
> Thanks again @sillent - it sounds like just switching to `FuturesOrdered` is the best way forward, given it seems like a strictly better default.

Cool, i switched back the all changes and do only one  - change FuturesUnordered to FuturesOrdered

**Review by bruceg — APPROVED** [review]
All checks are green and the change seems simple enough. I'm good with this if others are.

**Comment by spencergilbert** [comment]
Thanks again for the contribution @sillent!

**Pull request merged**